### PR TITLE
add SGD and Adam Optimizer for DeepES

### PR DESCRIPTION
**Pull request by rical730**
1. add SGDOptimizer with momentum computation supported 
2. add AdamOptimizer
3. switch Optimizer by deepes_config.prototxt
4. update deepes_config.prototxt

### CODE REVIEW

File: deepes/include/torch_predictor.h [hunk 1]
```diff
@@ -40,7 +42,16 @@ class Predictor{
     load_proto_conf(config_path, *_config);
     _sampling_method = std::make_shared<GaussianSampling>();
     _sampling_method->load_config(*_config);
-    _optimizer = std::make_shared<SGDOptimizer>(_config->optimizer().base_lr());
+
+    std::string opt_type = _config->optimizer().type();
+    std::transform(opt_type.begin(),opt_type.end(),opt_type.begin(),::tolower);
+    if (opt_type == "sgd") {
+      _optimizer = std::make_shared<SGDOptimizer>(_config->optimizer().base_lr());
```
Comment (TomorrowIsAnOtherDay) [inline deepes/include/torch_predictor.h]: Please add additional configuration term for these optimizers, like momentum for sgd and epsilon for `Adam` optimizer.

File: deepes/src/adam_optimizer.cc [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cmath>
+#include "adam_optimizer.h"
+
+namespace DeepES {
+
+AdamOptimizer::~AdamOptimizer() {
+	for (std::map<std::string, float*>::iterator iter = _momentum.begin(); iter != _momentum.end(); iter++) {
```
Comment (TomorrowIsAnOtherDay) [inline deepes/src/adam_optimizer.cc]: Please use `auto` for the iteration, a more readable way.

File: deepes/src/adam_optimizer.cc [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		delete[] iter->second;
+	}
+	for (std::map<std::string, float*>::iterator iter = _velocity.begin(); iter != _velocity.end(); iter++) {
```
Comment (TomorrowIsAnOtherDay) [inline deepes/src/adam_optimizer.cc]: Please use `auto` for the iteration, a more readable way.

File: deepes/src/sgd_optimizer.cc [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cmath>
+#include "sgd_optimizer.h"
+
+namespace DeepES {
+
+SGDOptimizer::~SGDOptimizer() {
+	for (std::map<std::string, float*>::iterator iter = _velocity.begin(); iter != _velocity.end(); iter++) {
```
Comment (TomorrowIsAnOtherDay) [inline deepes/src/sgd_optimizer.cc]: Please use `auto` for the iteration, a more readable way.

File: deepes/include/torch_predictor.h (hunk 1 above)
Comment (rical730) [reply deepes/include/torch_predictor.h]: done

File: deepes/src/adam_optimizer.cc (hunk 2 above)
Comment (rical730) [reply deepes/src/adam_optimizer.cc]: done

File: deepes/src/adam_optimizer.cc (hunk 3 above)
Comment (rical730) [reply deepes/src/adam_optimizer.cc]: done

File: deepes/src/sgd_optimizer.cc (hunk 4 above)
Comment (rical730) [reply deepes/src/sgd_optimizer.cc]: done

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
Looks good to me.

**Pull request merged**